### ISSUE TITLE
Using logstash collector 

### ISSUE THREAD

**Issue by Charlotte-Holt** [OpenLiberty/docs#1684]
New task topic to be moved over from the KC: https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_analytics_logstash.html. 

Confirm with SME that all info in the KC topic is still relevant/useful. 

**SME**: Don Bourne

**Comment by dmuelle** [comment]
Comments from @donbourne in #1095 

I would suggest taking the original topic and just making the following modifications to create the OL topic:

eliminate reference to Logstash V2.x, Elasticsearch V2.x, Kibana V4.x
eliminate reference to bluemixLogCollector
change "Enable HTTP access logging and use the '%h %u %t "%r" %s %b %D %{User-agent}i' log format." to "Enable HTTP access logging."

**Comment by dmuelle** [comment]
Hi @donbourne I have a draft of this topic, which is based on the [Using the Logstash collector version 1.0](https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_analytics_logstash.html) KC topic. I have a few questions. When you have a chance, can you review the draft and my questions below? If possible, can you get back to me by Friday 6/19? If there is anyone else I should reach out to instead, just let me know.

[Here's the initial draft](https://draft-openlibertyio.mybluemix.net/docs/ref/general/#collecting-logs-logstash.html)

## Title
I changed the title to "Collecting logs and events with Logstash collector" to give the reader more information than just "Using the Logstash collector..." Is this an ok title? Should the title reference Kibana or the Elastic stack? Most of the steps here move toward the goal of getting the log data intop a Kibana dashboard.

## About this task
- Is it out of the scope of this topic to advise the user of cases where they would want to use Logstash? I made mention that it can provide data to various analysis tools but that users preferred it for use with the ELK stack. Is there any other concise information we can include to tell the user why they should care about Logstash?
- Per your comment, I removed reference to "Logstash V2.x, Elasticsearch V2.x, Kibana V4.x". Should directions for KIbana 4 also be removed from steps 7, 9, &10?

## Procedure
### Step 1
"Set up Logstash by following the instructions from Elastic."
Should this be a prereq rather than the first step? Do users also need to set up Elastic in order to work through these steps? The sample repo references Elastic versions
### Step 2
"Create or acquire certificate and key pair files for SSL for Logstash" 
Should this step reference TLS as well? Do these files enable secure connections for Logstash? If so could rewrite as "Create or acquire certificate and key pair files to enable SSL or TLS connections for Logstash." Thoughts?
### Step 3
"For Logstash and Elasticsearch users..." 
Aren't all these steps for ELK users? This is a global question for this topic. Do we need to make more space for non-ELK uses for Logstash collector, or, is this use so common that we don't need to call it out?
### Step 3a
"Download the liberty_logstash.conf Logstash configuration file and the liberty_logstash_template.json index template file from the directory that corresponds to your Elasticsearch version"

The KC topic said
"... for your Elastic stack version according to the readme file."
But the readme file in that repo doesn't appear to give any details that are relevant here. It's just one sentence that points back to the KC topic. Does my edit give enough info?

### Step 4d
actual feature config example is in the logstash feature doc, per OL content guidelines. We can updated that config as needed to make sure it has the correct info
### Step 4e
"Enable HTTP access logging."
Where can we point users for more info on this? Is `audit-1.0` the feature that OL users need? or `monitor-1.0`?

Let me know what else you see that needs attention. Thanks!

**Comment by donbourne** [comment]
@dmuelle 

Your questions
- I'm wondering if title should be "Forwarding logs and events to Logstash with Logstash collector"?
- I don't think the title should reference Kibana or Elastic stack
- I think there's another topic in progress that will tell people when to use logstashCollector vs. JSON logging
- see my general comments (below) about why users should care about Logstash -- it has lots of good output plugins
- yes, probably a good idea to skip documenting Kibana 4 in steps 7,9,10
- yes, would probably be a good idea to move setting up logstash to prereq along with set up Elasticsearch and set up Kibana.  url should change to https://www.elastic.co/
- step 2 - "Create or acquire certificate and key pair files to enable SSL or TLS connections for Logstash." - looks good
- step 3 - yes, these steps are all for ELK users.  Might make more sense to put a comment, before we start to describe the procedure, to say that this is how to get set up with ELK, and that setting up to use Logstash with other outputs can be done similarly (exercise left to the reader?)
- step 3a - yes, your edits look good
- step 4d - sounds good
- step 4e - I would point people to https://openliberty.io/docs/ref/general/#access-logging.html

Other Comments
"by products such as Elasticsearch, Kibana, Kafka, and RabbitMQ" 
- these aren't all products related to log analysis.  we might want to say that Logstash can ingest and transform data then send it to a variety of outputs such as Kafka, an email server, Elasticsearch or PagerDuty.

- remove step 4c (Run the following command to install the logstashcollector-1.0 feature) since OL doesn't yet have a command for adding features

**Comment by dmuelle** [comment]
Thanks for reviewing @donbourne, addressed you comments in a new draft.

https://draft-openlibertyio.mybluemix.net/docs/ref/general/#forwarding-logs-logstash.html

 - title should be "Forwarding logs and events to Logstash with Logstash collector"? **done**
- skip documenting Kibana 4 in steps 7,9,10 **done**
- move setting up logstash to prereq along with set up Elasticsearch and set up Kibana.  url should change to https://www.elastic.co/ **done**
- step 3 - yes, these steps are all for ELK users.  Might make more sense to put a comment, before we start to describe the procedure, to say that this is how to get set up with ELK, and that setting up to use Logstash with other outputs can be done similarly (exercise left to the reader?) **done, included a link to Logstash output plugins for more info**
- step 4e - I would point people to https://openliberty.io/docs/ref/general/#access-logging.html **done**
- "by products such as Elasticsearch, Kibana, Kafka, and RabbitMQ" 
- these aren't all products related to log analysis.  we might want to say that Logstash can ingest and transform data then send it to a variety of outputs such as Kafka, an email server, Elasticsearch or PagerDuty. **done**
- remove step 4c (Run the following command to install the logstashcollector-1.0 feature) since OL doesn't yet have a command for adding features **done**

One question: in step 6 "For Kibana 5.0 - 5.5,..." the step does not say to specify an index pattern (`logstash-*`) like it does for the other versions- is this correct? Or do you need to specify the index pattern no matter what, and how you  select datetime as the Time filter field name is different for version 5.0 - 5.5?

Lmk know if any other edits are needed. If the topic looks good, can you add the `Technical Reviewed` label to this issue? Thanks

**Comment by donbourne** [comment]
"Together, these three tools comprise the Elastic stack"
"To configure the Elastic stack"
- Elastic stack is usually described as ELK + Beats.  It may not be worth getting into what the Elastic stack is in this doc, so I'd suggest just avoiding any term for that.
---
included a link to Logstash output plugins for more info
- https://www.elastic.co/guide/en/logstash/current/output-plug-ins.html gives me a 404.  May not be stable enough to try to link directly to their output plugin doc.  Perhaps just link again to elastic.co
---

send it to a Logstash server, and view it in a Kibana dashboard
->
send it to a Logstash server, store it in Elasticsearch, and view it in a Kibana dashboard

---
in last batch of comments I made incorrect statement about OL not having command to install the logstashCollector-1.0 feature.  It now has featureManager to do that, so I'd suggest we add the following before 3c:
```
Install Logstash collector feature if it's not already installed on your server
d:\wlp\bin\featureUtility installFeature logstashCollector-1.0
```
---
One question: in step 6 "For Kibana 5.0 - 5.5,..." the step does not say to specify an index pattern (logstash-*) like it does for the other versions- is this correct? Or do you need to specify the index pattern no matter what, and how you select datetime as the Time filter field name is different for version 5.0 - 5.5?
- I'm not sure -- I'm inclined to assume it's correct since we had tested these instructions before
---
You configured your Open Liberty servers to send events to your Logstash server and can now view your events in the Liberty dashboard by using Kibana.
->
You configured your Open Liberty servers to send events to your Logstash server and can now view your events in the Liberty dashboards by using Kibana.

**Comment by dmuelle** [comment]
Hi @donbourne - thanks for reviewing, I've addressed your comments in a new draft and also added some more info to clarify when a user would need to use Logstash collector vs when the container or managing agent would handle sending the logs to Logstash. Can you please review the updated **About this task** section and let me know if my explanation makes sense?

RE your previous comments:
- removed references to "the Elastic stack"
- I had a typo in that link to logstash output plugins, which I fixed. If possible, I'd like to keep this link because it's not easy to find this info starting from the Elastic homepage
- added the feature utility command substep to step 3
- re step 6- The information for Logstash 5.0-5.5 is consistent with what's in the other KC topic (Analyzing Logs with ELK stack)[https://www.ibm.com/support/knowledgecenter/en/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_elk_stack.html]. However, that topic says to use `ibm_datetime` as the time filter field name while this topic says to use `datetime`. I'm assuming this is intentional but thought it was worth checking.
- made "dashboards" plural in the Results sentence noted above

I will be creating a new topic in OL for Analyzing Logs with ELK stack that will be linked to this topic so that users can find the info they need depending on their configuration. IS nit possible tom have someone test both these procedures to validate the steps? 

Thanks

**Comment by donbourne** [comment]
Hi @dmuelle , this is looking really good.

For step 6, yes - `datetime` is correct for Logstash collector and it is different from the JSON we send when JSON logging is enabled.

such as a Kubernetes container -> such as a Kubernetes cluster

collect your log data at run time and send it to a Logstash server -> collect your log data at run time and send it directly to a Logstash server

#1903 needs a little work -- it's confusing a completely standalone environment, where you would set up your own ELK stack, with a Kubernetes environment, where all of that typically gets installed with Cluster Logging.

I think this topic is ready to test.  We should probably finish editing #1903 before testing that.  I'll see if someone on the observability squad can try this one out.

**Comment by dmuelle** [comment]
Hi @donbourne- thanks for reviewing- I've made the two edits you suggested.

- [x] collect your log data at run time and send it to a Logstash server -> collect your log data at run time and send it directly to a Logstash server

- [x] collect your log data at run time and send it to a Logstash server -> collect your log data at run time and send it directly to a Logstash server

Is it possible to get this topic tested this week? Would it be easier to wait until #1903 is ready, and have someone test them both?
re #1903 , lmk in that issue what edits I can make to clear up the confusion. Thanks!

**Comment by donbourne** [comment]
If your Open Liberty server is deployed in an environment that manages your logs, such as a Kubernetes cluster, the managing agent can transfer your log data to Logstash 
->
If your Open Liberty server is deployed in an environment that manages your logs, such as a Kubernetes cluster, your logs will typically be aggregated by agents running on the platform.

**Comment by donbourne** [comment]
> Is it possible to get this topic tested this week?

I've opened an issue internally to request help to get this tested.  Will see if anyone can work on this now.

**Comment by dmuelle** [comment]
@donbourne 
`If your Open Liberty server is deployed in an environment that manages your logs, such as a Kubernetes cluster, your logs will typically be aggregated by agents running on the platform`

What does "platform" refer to in this context? Is it the "environment that manages your logs, such as a Kubernetes cluster" ? Or the log analysis platform? I want to be careful about how "platform:" is used in this doc so it doesn't have different meaning in different places. 

Would it be correct to say 

`If your Open Liberty server is deployed in a managed environment, such as a Kubernetes cluster, your logs are typically aggregated by agents in that environment.`

(also had to edit slightly to reduce complexity)

**Comment by donbourne** [comment]
> If your Open Liberty server is deployed in a managed environment, such as a Kubernetes cluster, your logs are typically aggregated by agents in that environment.

That sounds good.

**Comment by jennifer-c** [comment]
Hi, I tested the steps with ELK 7.6.0 and I have a few minor comments:

- Step 2a: `liberty_logstash_template.json` -> `liberty_logstash_template_ibm.json`
- Step 2b: Mention that the Elasticsearch template path also needs to be changed to wherever the user saved the index template. e.g. 
"Customize the Elasticsearch template path with the `liberty_logstash_template_ibm.json index template file."
- Step 3: If possible, it'd be great to have an option to switch it to a Unix-style command (since it's formatted for Windows the slashes are wrong when copy-pasting) - but not a big deal overall

**Comment by dmuelle** [comment]
Thanks for testing @jennifer-c ! I've updated the doc per your comments

- [x] Step 2a: liberty_logstash_template.json -> liberty_logstash_template_ibm.json **fixed**
- [x] Step 2b: Mention that the Elasticsearch template path also needs to be changed **updated**
- [ ] Step 3: If possible, it'd be great to have an option to switch it to a Unix-style command (since it's formatted for Windows the slashes are wrong when copy-pasting)- **this would be a great feature, site wide. But we don't have that functionality at the moment. I'll ask about it on the next design call**

**Comment by dmuelle** [comment]
@donbourne I believe I've addressed all the issues you raised, as well as those from Jennifer's test of the doc. Are any more updates needed? If not, can you add the `technical reviewed` label to this issue? Thanks

**Comment by dmuelle** [comment]
new draft URL: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/forwarding-logs-logstash.html

**Comment by donbourne** [comment]
Looking at both this topic and #1903, I'm wondering if we could have them both use the substitution text below (obviously with the forwarding link in the 3rd paragraph when this text is used in #1903)...

**Change:**
If your Open Liberty server is deployed in a managed environment, such as a Kubernetes cluster, your logs are typically aggregated by agents in that environment. However, some server configurations cannot use an external agent to manage log data. In these cases, you can use the Open Liberty Logstash collector feature to internally collect your log data at run time and send it directly to a Logstash server.

If your log data is managed by an external agent, you do not need to use the Logstash collector feature. For more information, see Analyzing logs with the Elastic stack.

**To:**
If your Open Liberty server is deployed in a managed environment, such as a Kubernetes cluster, your logs are typically aggregated by agents in that environment.

If instead you are able to install Filebeat on the same host as your Open Liberty server, you can configure Open Liberty to write logs in JSON format and configure Filebeat to forward the logs. For more information, see Analyzing JSON logs with Elastic Stack.

In cases where your server logs are not managed and where it is not possible to install Filebeat on the same host as your Open Liberty server, you can use the Open Liberty Logstash collector feature to internally collect log data at run time and send it directly to a Logstash server.

**Comment by dmuelle** [comment]
Thanks for reviewing @donbourne - I've updated the doc per your comment, with slight modifications for IBM style. I also used this description in #1903, per your suggestion. If this needs anything else, let me know. If you're ready to sign off, please indicate in a comment and add the `techincal reviewed` label. Thanks!

**Comment by dmuelle** [comment]
https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/forwarding-logs-logstash.html

**Comment by donbourne** [comment]
looks good!

**Comment by dmuelle** [comment]
@Charlotte-Holt can you provide peer review

**Comment by dmuelle** [comment]
@lauracowen this is ready for your review- should be reviewed in tandem with #1903  as they are closely related

**Comment by lauracowen** [comment]
- "Logstash collector feature" -> "Logstash Collector feature"? (assuming the feature name is capitalised consistently...)
- Where it says "If your OL server is deployed in a managed environment..."
  - Might that agent also be Filebeat?
  - Is it worth having a second sentence saying something like "see the doc for that agent"? (sounds a bit rubbish so not sure)
  The reason I'm asking is that the three paragraphs read "another agent, Filebeat, Logstash Collector" but then the instructions on this page pick the middle one. I see why but it kind of made me initially wonder "why are you picking that one to explain?" - is that the one you prefer me to use? Obvs we can't explain how to use some unknown agent, I think it was just the rhythm threw me a little.
- "the instructions from Elastic." - is it possible to make that link point to their docs at least rather than their marketing homepage?
- Not sure you need the explanatory text before the links in See Also. I think they're good links to have there but I think the titles work ok. Where you want them to see the topic for a specific reason, you've linked it inline to the relevant section (re JSON logging) so I think it's fine without any more detail in See Also.
- Step 7 should be a numbered/ordered list nested under Kibana 7, 6, or 5.6 (not bullets).

**Comment by dmuelle** [comment]
Thanks @lauracowen - I've updated the draft

- "Logstash collector feature" -> "Logstash Collector feature"? **fixed**
- Where it says "If your OL server is deployed in a managed environment..."
**This is a bit tricky. Trying to cover three possible situations:**
1. your server is in a K8 cluster or similar. Your logging data is collected and sent to ELK at the cluster level, you don't need the Logstash Collector feature. The external agent _could_ be Filebeat (I think), but whether or not it is out of scope of this task.
2. you're not in a cluster, but you can install Filebeat on the same host as your server. Filebeat sends the logs to ELK. You don't need LC, but you might need the info in "Analyzing Logs with the Elastic Stack"
3. you're not in a cluster and you can't install filebeat on the same host as your server. You need the LC feature to collect and fwd your logs. This task is for you.

1 & 2 are related, but not the same. Does it make sense to mention them both? I think I started with something that said "You only need Logstash Collector if..." but in explaining the case, you kind of end up mentioning 1 & 2 anyway.

- "the instructions from Elastic." - is it possible to make that link point to their docs at least rather than their marketing homepage? **fixed**
- Not sure you need the explanatory text before the links in See Also. **removed**
- Step 7 should be a numbered/ordered list nested under Kibana 7, 6, or 5.6 (not bullets). **fixed**

**Comment by lauracowen** [comment]
Looks good, thanks. Signing off.

**Comment by dmuelle** [comment]
requesting peer review from @Charlotte-Holt

**Comment by Charlotte-Holt** [comment]
@dmuelle This looks good! Just a few comments:

## Peer review

- Global question (and possibly something for OL style guidelines) - Should sentences that lead into examples always end in a colon? Meaning should, "Customize the server name, password, and subject as needed." end in a colon? Or just sentences like, "Use the following command to install Logstash Collector feature if it is not already installed on your server:"?
- I think we're supposed to stay away from using links in the opening paragraph, per my review from Karen on the OpenShift topic. 

### Procedure

- "Customize the Beats ssl_certificate and ssl_key paths with the certificate and key pair files you generated in step 1."  -->  "Customize the Beats ssl_certificate and ssl_key paths with the certificate and key pair files that you generated in step 1."
- "Download a sample Logstash configuration file and an index template file from Sample Logstash Collector dashboards for Liberty."  -->  "Download a sample Logstash configuration file and an index template file from the Sample Logstash Collector dashboards for Liberty repository." (This is how it's referred to later in the topic, and I think it's a little clearer?)
- "Use the following command to install Logstash Collector feature if it is not already installed on your server:"  -->  "Use the following command to install the Logstash Collector feature if it is not already installed on your server:"

**Comment by dmuelle** [comment]
Thanks @Charlotte-Holt 
## Peer review fixes

- Global question (and possibly something for OL style guidelines) - Should sentences that lead into examples always end in a colon? Meaning should, "Customize the server name, password, and subject as needed." end in a colon? Or just sentences like, "Use the following command to install Logstash Collector feature if it is not already installed on your server:"?
**I made these statments end with a colon throughout for consistency. I think that's better than a case-by-case approach**
- I think we're supposed to stay away from using links in the opening paragraph, per my review from Karen on the OpenShift topic. **fixed**

### Procedure

- "Customize the Beats ssl_certificate and ssl_key paths with the certificate and key pair files you generated in step 1."  -->  "Customize the Beats ssl_certificate and ssl_key paths with the certificate and key pair files that you generated in step 1." **fixed**
- "Download a sample Logstash configuration file and an index template file from Sample Logstash Collector dashboards for Liberty."  -->  "Download a sample Logstash configuration file and an index template file from the Sample Logstash Collector dashboards for Liberty repository." **fixed**
- "Use the following command to install Logstash Collector feature if it is not already installed on your server:"  -->  "Use the following command to install the Logstash Collector feature if it is not already installed on your server:" **fixed**

**Comment by dmuelle** [comment]
https://draft-openlibertyio.mybluemix.net/docs/latest/forwarding-logs-logstash.html

**Comment by dmuelle** [comment]
Task structure and headings revised per updated guidance in #2656
Edits for clarity and concreteness in #3204
- reduce wordiness
-  Use specific language
- Consider the skill level and needs of users
- Set the context for examples and scenarios

Rechecked against the[ KC source topic ](https://www.ibm.com/support/knowledgecenter/SSEQTP_liberty/com.ibm.websphere.wlp.doc/ae/twlp_analytics_logstash.html)to make sure nothing was unintentionally omitted

**Comment by dmuelle** [comment]
Doc is now on vNext branch, will publish with 20.0.0.12 on 11.20

**Comment by dmuelle** [comment]
https://openliberty.io/docs/latest/forwarding-logs-logstash.html
